pkg/controllers/management/auth/globalroles/globalrole_handler.go

```diff
@@ -196,6 +217,134 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
+
+		if namespace.Status.Phase == corev1.NamespaceTerminating {
+			// If the namespace is terminating, don't try to make or update Roles
+			logrus.Warnf("[%v] Namespace %s is terminating. Not creating roles for %s", grController, ns, globalRole.Name)
+			continue
+		}
+
+		// Check if the role exists
+		roleName := wrangler.SafeConcatName(globalRole.Name, ns)
+		role, err := gr.rLister.Get(ns, roleName)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+			newRole := &v1.Role{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      roleName,
+					Namespace: ns,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: globalRole.APIVersion,
+							Kind:       globalRole.Kind,
+							Name:       globalRole.Name,
+							UID:        globalRole.UID,
+						},
+					},
+					Labels: map[string]string{
+						grOwnerLabel: globalRoleName,
+					},
+				},
+				Rules: rules,
+			}
+			createdRole, err := gr.rClient.Create(newRole)
+			if err == nil {
+				roleUIDs[createdRole.UID] = struct{}{}
+				continue
+			}
+
+			if !apierrors.IsAlreadyExists(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+
+			// In the case that the role already exists, we get it and check that the rules are correct
+			role, err = gr.rLister.Get(ns, roleName)
+			if err != nil {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+		}
+		if role != nil {
+			roleUIDs[role.GetUID()] = struct{}{}
+
+			// Check that the rules for the existing role are correct and that it has the right Owner Label
+			if reflect.DeepEqual(role.Rules, rules) && role.Labels != nil && role.Labels[grOwnerLabel] == globalRoleName {
+				continue
+			}
+
+			newRole := role.DeepCopy()
+			newRole.Rules = rules
+			if newRole.Labels == nil {
+				newRole.Labels = map[string]string{}
+			}
+			newRole.Labels[grOwnerLabel] = globalRoleName
+
+			_, err := gr.rClient.Update(newRole)
+			if err != nil {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+		}
+	}
+
+	// get all the roles claiming to be owned by this GR and remove any that shouldn't exist
+	r, err := labels.NewRequirement(grOwnerLabel, selection.Equals, []string{globalRoleName})
+	if err != nil {
+		return multierror.Append(returnError, errors.Wrapf(err, "couldn't create label: %s", grOwnerLabel))
+	}
+
+	roles, err := gr.rLister.List("", labels.NewSelector().Add(*r))
+	if err != nil {
+		return multierror.Append(returnError, errors.Wrapf(err, "couldn't list roles with label %s : %s", grOwnerLabel, globalRoleName))
+	}
+
+	// After creating/updating all Roles, if the number of RBs with the grOwnerLabel is the same as
+	// as the number of created/updated Roles, we know there are no invalid Roles to purge
+	if len(roleUIDs) != len(roles) {
```

Comment:
Is this comment true? I'm thinking that there could be a case where a valid role was removed and an invalid role was added - in these cases, you would not purge the invalid Roles.

I don't think that this is too concerning since in theory the add/remove should trigger a re-enqueue, which should create the new role, and then trigger the removal here, but I thought it would at least be worth mentioning.

Reply:
In order to cause that to happen, the Role would have to be deleted and an invalid Role be created after the original Role was added to the UID map in the above for loop but before this check happens. While possible, I think it's very unlikely. And as you said, creating an invalid Role with the `authz.management.cattle.io/gr-owner` label set to an existing GR triggers another reconcile via `roleEnqueueGR` and that reconcile would fix it.

Would you like me to mention that edge case in the comment? Ditto for GRB

Reply:
I believe in that case, I should remove this `if` and let the purge function run on all Roles found from the `list`.

The alternative I can think of is doing a `list` at the beginning, turning the slice of Roles into a map and then as I go through the NamespacedRules, removing them from the map as I create them. Then at the end I would have a list of Roles that shouldn't be there. This has the same above problem though where new Roles could be created after the `List` call. It also isn't any more efficient since I'd have to loop through the slice of Roles to create a map.

By doing the `List` at the very end, I get the most up to date list of Roles and purge any that are invalid. Let me know what you think, or if you have any other suggestions.